### PR TITLE
Add support for interpolated strings ("interstrings")

### PR DESCRIPTION
**Pull request by KevinEady**
## Interpolated string expression: `$"..."`

This new expression syntax allows you to join multiple components into a string. Within the `...`, you may specify zero or more:
- normal strings
- escaped characters, eg. `\n`
- `{expression [: format]}`, eg. `{who.name}` or `{who.graphic : #x}`
- `{{` for a `{` character, and `}}` for a `}` character

A complete example:
```cs
$"Hello, {who.name}! You are player #{who in EnumerateOnlineCharacters()} online. Your graphic is {who.graphic : #x}."
```
... can replace an expression like ...
```cs
"Hello " + who.name + "! You are player #" + (who in EnumerateOnlineCharacters()) + " online. Your graphic is " + Hex(who.graphic)
```
... or ...
```cs
"Hello {1}! You are player #{2} online. Your graphic is {3:#x}".format(who.name, who in EnumerateOnlineCharacters(), who.graphic)
```
Since this change introduces new instructions, you **MUST** recompile all your scripts!

### CODE REVIEW

**Comment by KevinEady** [comment]
Discord conversations for tracking:
- Interstrings as struct initializers
  - https://discord.com/channels/332123400573026306/657361447998849074/813515714086436875

File: lib/EscriptGrammar/EscriptLexer.g4 [hunk 1]
```diff
@@ -1,6 +1,12 @@
 lexer grammar EscriptLexer;
 channels { COMMENTS }
 
+@lexer::members
```
Comment (KevinEady) [inline lib/EscriptGrammar/EscriptLexer.g4:4]: The lexer updates are primarily based off of C# here: https://github.com/antlr/grammars-v4/blob/master/csharp/CSharpLexer.g4 , minus the use of "verbatim string"

File: lib/EscriptGrammar/EscriptParser.g4 [hunk 2]
```diff
@@ -377,6 +378,17 @@ literal
     | STRING_LITERAL
     ;
 
+interpolatedString
```
Comment (KevinEady) [inline lib/EscriptGrammar/EscriptParser.g4:381]: The parser updates are primarily based off C# here: https://github.com/antlr/grammars-v4/blob/master/csharp/CSharpParser.g4 , minus the use of expression lists

File: pol-core/bscript/compiler/ast/FormattedString.cpp [hunk 3]
```diff
@@ -0,0 +1,31 @@
+#include "FormattedString.h"
```
Comment (KevinEady) [inline pol-core/bscript/compiler/ast/FormattedString.cpp:1]: Is this the correct name for this AST class?

File: pol-core/bscript/compiler/ast/FormattedString.cpp [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+void FormattedString::accept( NodeVisitor& visitor )
+{
+  visitor.visit_formatted_string( *this );
+}
+
+void FormattedString::describe_to( fmt::Writer& w ) const
+{
+  w << "format-as-string";
```
Comment (KevinEady) [inline pol-core/bscript/compiler/ast/FormattedString.cpp:28]: Is this the correct description of the AST node...?

File: pol-core/bscript/compiler/ast/FormattedString.cpp [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+}  // namespace Pol::Bscript::Compiler
```
Comment (KevinEady) [inline pol-core/bscript/compiler/ast/FormattedString.cpp:31]: \n

File: pol-core/bscript/compiler/ast/FormattedString.cpp [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+#include <format/format.h>
+
+#include "clib/strutil.h"
+#include "compiler/ast/NodeVisitor.h"
+#include "compiler/ast/StringValue.h"
+
+namespace Pol::Bscript::Compiler
+{
+FormattedString::FormattedString( const SourceLocation& source_location,
```
Comment (KevinEady) [inline pol-core/bscript/compiler/ast/FormattedString.cpp:11]: A formatted string has two children: the format and the expression

File: pol-core/bscript/compiler/ast/InterpolatedString.cpp [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+#include <format/format.h>
+
+#include "clib/strutil.h"
+#include "compiler/ast/NodeVisitor.h"
+
+namespace Pol::Bscript::Compiler
+{
+InterpolatedString::InterpolatedString( const SourceLocation& source_location,
+                                        std::vector<std::unique_ptr<Expression>> children )
```
Comment (KevinEady) [inline pol-core/bscript/compiler/ast/InterpolatedString.cpp:11]: An interpolated string has `n` expression children, so a vector of expressions

File: pol-core/bscript/compiler/ast/InterpolatedString.cpp [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+{
+  visitor.visit_interpolated_string( *this );
+}
+
+void InterpolatedString::describe_to( fmt::Writer& w ) const
+{
+  w << "interpolated-string (" << children.size() << " parts)";
+}
+
+}  // namespace Pol::Bscript::Compiler
```
Comment (KevinEady) [inline pol-core/bscript/compiler/ast/InterpolatedString.cpp:26]: \n

File: pol-core/bscript/compiler/ast/InterpolatedString.cpp [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+void InterpolatedString::accept( NodeVisitor& visitor )
+{
+  visitor.visit_interpolated_string( *this );
+}
+
+void InterpolatedString::describe_to( fmt::Writer& w ) const
+{
+  w << "interpolated-string (" << children.size() << " parts)";
```
Comment (KevinEady) [inline pol-core/bscript/compiler/ast/InterpolatedString.cpp:23]: Is this the correct description of the AST node...?

File: pol-core/bscript/compiler/astbuilder/ExpressionBuilder.cpp [hunk 10]
```diff
@@ -274,6 +290,41 @@ std::vector<std::unique_ptr<Expression>> ExpressionBuilder::expressions(
   return expressions;
 }
 
+std::vector<std::unique_ptr<Expression>> ExpressionBuilder::expressions(
+    std::vector<EscriptGrammar::EscriptParser::InterpolatedStringPartContext*> ctx )
```
Comment (KevinEady) [inline pol-core/bscript/compiler/astbuilder/ExpressionBuilder.cpp:294]: This is the function that takes the Parse Tree node `InterpolatedStringPartContext` and returns the list of its children expressions, ie. `vector<Expression>`

File: pol-core/bscript/compiler/astbuilder/ExpressionBuilder.cpp [hunk 11]
```diff
@@ -274,6 +290,41 @@ std::vector<std::unique_ptr<Expression>> ExpressionBuilder::expressions(
...
+  // FIXME: optimize values
+  // FIXME: double-curleys
+  for ( auto interstringPart_ctx : ctx )
+  {
+    if ( auto expression_ctx = interstringPart_ctx->expression() )
+    {
+      std::unique_ptr<Expression> expr = expression( expression_ctx );
+      if ( auto format = interstringPart_ctx->FORMAT_STRING() )
+      {
+        expr = formatted_string( std::move( expr ), format );
```
Comment (KevinEady) [inline pol-core/bscript/compiler/astbuilder/ExpressionBuilder.cpp:306]: If there's a format string, create a new `FormattedString` ast node, using the interstring expression as its child.

File: pol-core/bscript/compiler/astbuilder/ValueBuilder.cpp [hunk 12]
```diff
@@ -62,22 +62,22 @@ std::unique_ptr<IntegerValue> ValueBuilder::integer_value(
...
-    antlr4::tree::TerminalNode* string_literal )
+    antlr4::tree::TerminalNode* string_literal, bool expect_quotes )
 {
   auto loc = location_for( *string_literal );
-  return std::make_unique<StringValue>( loc, unquote( string_literal ) );
+  return std::make_unique<StringValue>( loc, unquote( string_literal, expect_quotes ) );
 }
 
-std::string ValueBuilder::unquote( antlr4::tree::TerminalNode* string_literal )
+std::string ValueBuilder::unquote( antlr4::tree::TerminalNode* string_literal, bool expect_quotes )
```
Comment (KevinEady) [inline pol-core/bscript/compiler/astbuilder/ValueBuilder.cpp:71]: When using this in a normal string, the start and ending tokens are `"`. This is not the case with interstrings: `$"Hello!"` -> the `$"` is interstring_start, and the `H` begins the string part

File: pol-core/bscript/compiler/codegen/InstructionEmitter.cpp [hunk 13]
```diff
@@ -386,6 +386,16 @@ void InstructionEmitter::value( const std::string& v )
...
 
+void InstructionEmitter::interpolated_string( unsigned count )
+{
+  emit_token( TOK_INTERPOLATED_STRING, TYP_OPERAND, count );
+}
+
+void InstructionEmitter::formatted_string()
+{
+  emit_token( TOK_FORMATTED_STRING, TYP_OPERAND );
+}
```
Comment (KevinEady) [inline pol-core/bscript/compiler/codegen/InstructionEmitter.cpp:397]: Correct ... operand?

File: pol-core/bscript/compiler/format/StoredTokenDecoder.cpp [hunk 14]
```diff
@@ -386,6 +386,12 @@ void StoredTokenDecoder::decode_to( const StoredToken& tkn, fmt::Writer& w )
     w << "peek at top of stack; skip " << tkn.offset
       << " instructions if true, otherwise consume it";
     break;
+  case TOK_INTERPOLATED_STRING:
+    w << "interpolated string join (" << tkn.offset << " parts)";
+    break;
+  case TOK_FORMATTED_STRING:
+    w << "formatted string";
```
Comment (KevinEady) [inline pol-core/bscript/compiler/format/StoredTokenDecoder.cpp:393]: These are correct as well..?

File: pol-core/bscript/executor.cpp [hunk 15]
```diff
@@ -1258,6 +1260,41 @@ void Executor::ins_jmpiffalse( const Instruction& ins )
...
+    std::vector<std::string> contents( count );
+    while ( count-- )
+    {
+      BObjectRef rightref = ValueStack.back();
+      ValueStack.pop_back();
+      contents[count] = rightref->impptr()->getStringRep();
+    }
+    auto joined = std::accumulate( std::next( contents.begin() ), contents.end(), contents[0],
+                                   []( std::string a, std::string b ) { return a + b; } );
+    ValueStack.push_back( BObjectRef( new BObject( new String( joined ) ) ) );
```
Comment (KevinEady) [inline pol-core/bscript/executor.cpp:1281]: May be better way to do this?

File: pol-core/bscript/str.cpp [hunk 16]
```diff
@@ -890,6 +891,12 @@ bool try_to_format( std::stringstream& to_stream, BObjectImp* what, std::string&
 }
 // --
 
+std::string try_to_format(BObjectImp* what, std::string& frmt) {
```
Comment (KevinEady) [inline pol-core/bscript/str.cpp:894]: Maybe a different name

File: pol-core/bscript/tkn_strm.cpp [hunk 17]
```diff
@@ -469,6 +469,9 @@ void Token::printOn( std::ostream& os ) const
   case RSV_ELVIS:
     os << "?: (elvis)";
     break;
+  case TOK_INTERPOLATED_STRING:
```
Comment (KevinEady) [inline pol-core/bscript/tkn_strm.cpp:472]: missing formatted string

**Review by KevinEady — COMMENTED** [review]
A lot of these comments can be moved over to the code actually...

File: pol-core/bscript/executor.cpp (hunk 15 above)
Comment (KevinEady) [reply pol-core/bscript/executor.cpp:1281]: so something like this:
```cpp
std::string joined;
while ( count-- )
{
 BObjectRef rightref = ValueStack.back();
 ValueStack.pop_back();
 joined += rightref->impptr()->getStringRep();
}
```

File: pol-core/bscript/str.cpp (hunk 16 above)
Comment (KevinEady) [reply pol-core/bscript/str.cpp:894]: Suggest `get_formatted()`?

File: pol-core/bscript/compiler/ast/FormattedString.cpp (hunk 3 above)
Comment (KevinEady) [reply pol-core/bscript/compiler/ast/FormattedString.cpp:1]: We will use name `FormatExpression`.

File: pol-core/bscript/compiler/ast/FormattedString.cpp (hunk 4 above)
Comment (KevinEady) [reply pol-core/bscript/compiler/ast/FormattedString.cpp:28]: We will use `format-expression`

**Comment by KevinEady** [comment]
image
image
image

**Review by frozenblit — APPROVED** [review]

**Pull request merged**